File: galite-core/src/main/java/org/kopi/vkopi/lib/ui/swing/form/DImageField.java
/*
 * Copyright (c) 1990-2021 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id: DImageField.java 34961 2016-11-04 17:20:49Z hacheni $
 */

package org.kopi.vkopi.lib.ui.swing.form;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.Image;
import java.awt.event.ActionEvent;
import java.awt.event.KeyEvent;

import javax.swing.AbstractAction;
import javax.swing.ImageIcon;
import javax.swing.JComponent;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.KeyStroke;

import org.kopi.galite.visual.form.VFieldUI;
import org.kopi.galite.visual.form.VImageField;

/**
 * DImageField is a panel composed in a Image field and a label behind
 */
public class DImageField extends DObjectField {

  // ----------------------------------------------------------------------
  // CONSTRUCTION
  // ----------------------------------------------------------------------


/**
   * Constructor
   *
   * @param	model		the model for this text field
   * @param	label		The label that describe this field
   * @param	height		the number of line
   * @param	width		the number of column
   * @param	options		The possible options (NO EDIT, NO ECHO)
   */
  public DImageField(VFieldUI model,
		     DLabel label,
		     int align,
		     int options,
		     int width,
		     int height,
                     boolean detail)
  {
    super(model, label, align, options, detail);
    this.width = width;
    this.height = height;

    icon = new JLabel();

    empty = new JPanel() {
      /**
       * Comment for <code>serialVersionUID</code>
       */
        private static final long serialVersionUID = 4603093785481453119L;

	public Dimension getPreferredSize() {
          return new Dimension(DImageField.this.width, DImageField.this.height);
        }
      };

    //    empty.setBackground(DObject.CLR_FLD_BACK);

    add(empty, BorderLayout.CENTER);

    registerKeyboardAction(new AbstractAction() {
        /**
         * Comment for <code>serialVersionUID</code>
         */
        private static final long serialVersionUID = -555322973317509685L;

	public void actionPerformed(ActionEvent e) {
          setObject(null);
        }},
      null,
      KeyStroke.getKeyStroke(KeyEvent.VK_BACK_SPACE, 0),
      JComponent.WHEN_FOCUSED);
  }

  // ----------------------------------------------------------------------
  // IMPLEMENTATION OF ABSTRACTS METHODS
  // ----------------------------------------------------------------------

  /**
   * Returns the object associed to record r
   *
   * @return	the displayed value at this position
   */
  public Object getObject() {
    return image;
  }

  /**
   * Sets the object associed to record r
   *
   * @param	s		the object to set in
   */
  public void setObject(Object s) {
    if (image == null) {
      remove(empty);
    } else {
      remove(icon);
    }

    image = (byte[])s;

    if (image != null) {
      icon.setIcon(new ImageIcon(new ImageIcon(image).getImage().getScaledInstance(width, height, Image.SCALE_FAST)));
    }

    if (image == null) {
      add(empty, BorderLayout.CENTER);
    } else {
      add(icon, BorderLayout.CENTER);
    }

    setBlink(false);
    setBlink(true);
    repaint();
  }

  // ----------------------------------------------------------------------
  // UI MANAGEMENT
  // ----------------------------------------------------------------------

  public void setDisplayProperties() {
    repaint();
  }

  // ----------------------------------------------------------------------
  // DRAWING
  // ----------------------------------------------------------------------

  /**
   * This method is called after an action of the user, object should
   * be redisplayed accordingly to changes.
   */

  public void updateAccess() {
    label.update(getModel(), getPosition());
  }

  public void updateText() {
    setObject(((VImageField)getModel()).getImage(model.getBlockView().getRecordFromDisplayLine(getPosition())));
    super.updateText();
  }

  public void updateFocus() {
    label.update(getModel(), getPosition());
    fireMouseHasChanged();
    super.updateFocus();
  }
  
  public void updateColor() {}

  /**
   * set blink state
   */
  public void setBlink(boolean start) {}

  // ----------------------------------------------------------------------
  // DATA MEMBERS
  // ----------------------------------------------------------------------

  private	JLabel			icon;		// the image component
  private	JPanel			empty;		// the no image component
  private	int			width;
  private	int			height;
  private	byte[]			image;
  /**
   * Comment for <code>serialVersionUID</code>
   */
  private static final long serialVersionUID = -4798732518710333986L;
}

File: galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/form/DImageField.kt
/*
 * Copyright (c) 2013-2022 kopiLeft Services SARL, Tunis TN
 * Copyright (c) 1990-2022 kopiRight Managed Solutions GmbH, Wien AT
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.kopi.galite.visual.ui.vaadin.form

import java.io.ByteArrayInputStream
import java.io.InputStream
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

import org.kopi.galite.visual.form.VFieldUI
import org.kopi.galite.visual.form.VImageField
import org.kopi.galite.visual.ui.vaadin.base.BackgroundThreadHandler
import org.kopi.galite.visual.ui.vaadin.base.BackgroundThreadHandler.access
import org.kopi.galite.visual.ui.vaadin.field.ImageField

import com.vaadin.flow.component.upload.FailedEvent
import com.vaadin.flow.component.upload.SucceededEvent
import com.vaadin.flow.component.upload.receivers.MemoryBuffer
import com.vaadin.flow.dom.DomEvent
import com.vaadin.flow.server.StreamResource

/**
 * The image field implementation.
 *
 * @param model The field model.
 * @param label The field label.
 * @param align The field alignment.
 * @param options The field options
 * @param width The field width.
 * @param height The field height.
 * @param detail Does the field belongs to detail view ?
 */
class DImageField(
        model: VFieldUI,
        label: DLabel?,
        align: Int,
        options: Int,
        width: Int,
        height: Int,
        detail: Boolean
) : DObjectField(model, label, align, options, detail) {

  //---------------------------------------------------
  // DATA MEMBERS
  //---------------------------------------------------
  private var image: ByteArray? = null
  private var buffer = MemoryBuffer()
  private val field = ImageField(width.toFloat(), height.toFloat(), buffer)
  // --------------------------------------------------
  // CONSTRUCTION
  // --------------------------------------------------
  /**
   * Creates a new `DImageField` instance.
   */
  init {
    field.imageWidth = width
    field.imageHeight = height
    field.addObjectFieldListener(this)
    field.upload.addSucceededListener(::onUploadSucceeded)
    field.upload.addFailedListener(::onUploadFailed)
    field.upload.element.addEventListener("upload-abort", ::onRemove)
    setFieldContent(field)
  }

  // --------------------------------------------------
  // IMPLEMENTATION OF ABSTRACTS METHODS
  // --------------------------------------------------
  override fun getObject(): Any? = image

  /**
   * Sets the object associated to current record
   *
   * @param s The object to set in
   */
  fun setObject(s: ByteArray?) {
    access(currentUI) {
      if (s == null) {
        field.setData(s)
      } else {
        field.setData(DynamicImageResource(createFileName("image")) {
          ByteArrayInputStream(s as ByteArray?)
        })
        setBlink(false)
        setBlink(true)
      }
    }
    image = s
  }

  override fun setBlink(blink: Boolean) {
    // TODO
  }

  // ----------------------------------------------------------------------
  // UI MANAGEMENT
  // ----------------------------------------------------------------------
  override fun updateAccess() {
    label!!.update(model, position)
  }

  override fun updateText() {
    setObject((getModel() as VImageField).getImage(model.blockView.getRecordFromDisplayLine(position)))
    super.updateText()
  }

  override fun updateFocus() {
    label!!.update(model, position)
    super.updateFocus()
  }

  override fun updateColor() {
    // color properties are not set for an image field.
  }

  fun onRemove(event: DomEvent) {
    setObject(null)
    setImage(null)
  }

  /*fun onImageClick() { TODO
    performAutoFillAction()
  }*/

  /**
   * Creates the dynamic image name.
   * @param baseName The base name.
   * @return The dynamic image name.
   */
  private fun createFileName(baseName: String): String =
          baseName + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMddHHmmssSSS")) + ".png"
  //---------------------------------------------------
  // STREAM RESOURCE
  //---------------------------------------------------

  /**
   * A dynamic [StreamResource] for an image field.
   *
   * @param streamSource Returns an [InputStream] object.
   * @param fileName The file name.
   */
  internal class DynamicImageResource(fileName: String,
                                      streamSource: () -> InputStream)
    : StreamResource(fileName, streamSource) {

    init {
      cacheTime = 0L
    }
  }

  private fun onUploadSucceeded(event: SucceededEvent) {
    try {
      val bytes = buffer.inputStream.readBytes()
      setObject(bytes)
      setImage(bytes)
    } finally {
      if (event.contentLength > 50 * 1024 * 1024) {
        getModel().getForm().unsetProgressDialog()
      }
    }
  }

  fun setImage(bytes: ByteArray?) {
    getModel().isChangedUI = true
    getModel().setImage(bytes)
    getModel().onAfterDrop()
  }

  private fun onUploadFailed(event: FailedEvent) {
    event.reason.printStackTrace(System.err)
    Thread {
      getModel().getForm().error(event.reason.message)
      BackgroundThreadHandler.updateUI(currentUI)
    }.start()
  }
}
